### ISSUE TITLE
HaveFriend and REST HTTP Headers problem

### ISSUE THREAD

**Issue by Bobbe** [Codeception/Codeception#4203]
I'm trying to test multi session features.
I have 2 users. They are bearer authenticated with their tokens.
But when user2 did his part of job, user1 see token that belongs to user2.

In example I used header foo, just for problem illustration.
In real case when you use $I->amBearerAuthenticated($token) (what is actually just simple header like `foo` in example) the problem has more sense.

```bash
php codecept.phar run -vvv messages   
Codeception PHP Testing Framework v2.2.10
Powered by PHPUnit 5.7.17 by Sebastian Bergmann and contributors.

  Rebuilding MessageTester...

Messages Tests (1) ------------------
Modules: \Helper\User, Sequence, Asserts, \Helper\Config, REST, PhpBrowser
-------------------------------------

ExampleCest: Example
Signature: ExampleCest:example
Test: tests/messages/ExampleCest.php:example
Scenario --
 I have http header "foo","bar"
 USER2 DOES ---
 I have http header "foo","baz"
 I send get "/user"
  [Request] GET http://example.com/user
  [Request Headers] {"Foo":"baz"}
  [Page] http://example.com/user
  ...
 --- USER2 FINISHED
 I send get "/user"
  [Request] GET http://example.com/user
  [Request Headers] {"Foo":"baz"}
  [Page] http://example.com/user
  ...
```

Here is example test code that reproduce the problem
```php
class ExampleCest
{
    public function example(MessageTester $I)
    {
        $I->haveHttpHeader('foo', 'bar');
        $user2 = $I->haveFriend('user2', MessageTester::class);
        $user2->does(function(MessageTester $J) {
            $J->haveHttpHeader('foo', 'baz');
            $J->sendGET('/user');
        });
        //here you can see that header foo is baz
        $I->sendGET('/user');
    }
}
```

In conclusion, I can explain this behavior as single instance of REST module. 
But I think it's good idea to have separate instance of REST when you create a friend.

Current solution is to store token1 and token2 and always when you switch users use code 
`$I->amBearerAuthenticated(<token1 or token2>)`

**Comment by Naktibalda** [comment]
It should be enough to backup `headers` property in `_backupSession` method: https://github.com/Codeception/Codeception/blob/2.2.10/src/Codeception/Module/PhpBrowser.php#L267-L274

@Bobbe If you would like to do it yourself, please add a test to https://github.com/Codeception/Codeception/blob/2.2/tests/unit/Codeception/Module/PhpBrowserRestTest.php

**Comment by DavertMik** [comment]
Thanks, @Naktibalda 
I fixed it following your advice :+1:

**Comment by Naktibalda** [comment]
Thanks

**Comment by Naktibalda** [comment]
@DavertMik Please raise a pull request.

**Comment by DavertMik** [comment]
Yep, I will do it today

10 трав. 2017 р. 13:00 "Gintautas Miselis" <notifications@github.com> пише:
